### PR TITLE
[New features] MoE: add AllGather + ReduceScatter dispatcher with intermediate-dim EP sharding

### PR DESCRIPTION
**Pull request by adam-xiaoyao**
### PR Category

Distributed Strategy

### PR Types

New features

### Description

Introduce a new MoE token dispatcher — allgather — alongside the existing
deepep / alltoall / hybridep options. Semantic difference:

Existing dispatchers: each EP rank holds a subset of experts at full intermediate width
AllGather dispatcher: each EP rank holds every expert, with the intermediate dim sharded across EP

### 是否引起精度变化

否

### CODE REVIEW

File: src/paddlefleet/transformer/moe/token_dispatcher.py [hunk 1]
```diff
@@ -1114,5 +1142,675 @@ def combine_postprocess(self, permutated_local_input_tokens: paddle.Tensor):
...
+    Instead we build a fixed-size one-hot histogram: padding (-1) is sent to a
+    sink column ``== num_experts`` so it is dropped, and every other value is in
+    ``[0, num_experts)``.  All shapes are known a priori, so no host sync occurs.
+    Numerically identical to ``bincount(masked_select(indices, indices>=0),
+    minlength=num_experts)``.
+    """
+    flat = indices.reshape([-1]).cast("int64")
+    sink = paddle.full_like(flat, num_experts)
+    clamped = paddle.where(flat >= 0, flat, sink)
+    onehot = paddle.nn.functional.one_hot(clamped, num_classes=num_experts + 1)
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/token_dispatcher.py]: OK, 我改成scatter(overwrite=False)， 这里的显存可降两个数量级。

File: tests/single_card_tests/ai_edited_test/moe/test_ai_allgather_dispatcher.py [hunk 2]
```diff
@@ -0,0 +1,927 @@
...
+  - AllToAllTokenDispatcher.get_dispatched_routing / token_combine
+  - MoEFlexTokenDispatcher.get_dispatched_routing
+  - MoELayer._validate_allgather_config / _project_to_latent /
+    _maybe_pre_allgather_overlap / combine
+  - TransformerLayerWithOverlap ValueError for non-deepep dispatchers
+"""
+import os
+import sys
+
+sys.path.insert(
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply tests/single_card_tests/ai_edited_test/moe/test_ai_allgather_dispatcher.py]: OK, Remove unnecessary sys.path.insert hack in single-card test

File: tests/multi_card_tests/moe/test_allgather_dispatcher_ep.py [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+        cls.pg_collection = _ensure_fleet()
+
+    def setUp(self):
+        self.seed = 42
+        random.seed(self.seed)
+        np.random.seed(self.seed)
+        paddle.seed(self.seed)
+        paddle.manual_seed(self.seed)
+        model_parallel_cuda_manual_seed(self.seed)
+        self.ep_group = self.__class__.pg_collection.expert_parallel
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply tests/multi_card_tests/moe/test_allgather_dispatcher_ep.py]: OK, Fix ProcessGroupCollection field name (expert_parallel -> ep)

File: tests/single_card_tests/ai_edited_test/moe/test_ai_allgather_dispatcher.py [hunk 4]
```diff
@@ -0,0 +1,988 @@
...
+    """Test moe_allgather_gate_overlap config field exists with correct default."""
+
+    def test_default_false(self):
+        from paddlefleet.transformer.transformer_config import TransformerConfig
+
+        config = TransformerConfig()
+        self.assertFalse(config.moe_allgather_gate_overlap)
+
+
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply tests/single_card_tests/ai_edited_test/moe/test_ai_allgather_dispatcher.py]: OK, 我修改

File: tests/multi_card_tests/moe/test_allgather_dispatcher_ep.py [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+
+        layer = MagicMock()
+        layer.moe_token_dispatcher_type = "alltoall"
+        layer.token_dispatcher = MagicMock()
+        MoELayer.combine(layer, paddle.randn([4, 8]))
+        layer.token_dispatcher.token_combine.assert_called_once()
+        layer.token_dispatcher.combine_postprocess.assert_called_once()
+
+
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply tests/multi_card_tests/moe/test_allgather_dispatcher_ep.py]: OK ， 我修改。

File: src/paddlefleet/transformer/moe/token_dispatcher.py [hunk 6]
```diff
@@ -1114,5 +1140,725 @@ def combine_postprocess(self, permutated_local_input_tokens: paddle.Tensor):
...
+        grad_input = ReduceScatterGroupOp.apply(grad_output, group)
+        return grad_input
+
+
+def _reduce_scatter_async(input, group):
+    """Async ReduceScatter (SUM, axis 0) on the comm stream.
+    Returns (output, task)."""
+    input = input.contiguous()
+    out_shape = list(input.shape)
+    assert out_shape[0] % group.nranks == 0, (
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/token_dispatcher.py]: OK, 我修改。

File: tests/single_card_tests/ai_edited_test/moe/test_latent_moe.py [hunk 7]
```diff
@@ -833,6 +833,11 @@ def _make_stub(self, use_latent_moe, hidden_size=64, latent_size=32):
         stub.token_dispatcher._comm_manager.combine.return_value = paddle.randn(
             [bs_seq, expert_out_size]
         )
+        stub.token_dispatcher.get_dispatched_routing.return_value = (
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply tests/single_card_tests/ai_edited_test/moe/test_latent_moe.py:844]: OK， 我修改。

File: src/paddlefleet/transformer/moe/token_dispatcher.py [hunk 8]
```diff
@@ -1114,5 +1140,726 @@ def combine_postprocess(self, permutated_local_input_tokens: paddle.Tensor):
...
+    single uint8 row ``[T_local, H + 4*H128]`` (fp8 data ++ scale bytes).
+
+    Returns ``(fused_local, H, H128, scale_dtype)``. The caller is responsible
+    for AllGather (sync or async) and for unpacking via
+    :func:`_split_fused_fp8_gather`. Bit-identical to two separate gather of
+    data and scale: AllGather is a lossless row concat and the 1x128 scale
+    lives within a single token's hidden vector, so packing along axis 1
+    before the gather yields the same per-rank bytes as gathering separately.
+    """
+    assert quantize_activation_blockscaled_fast is not None, (
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/token_dispatcher.py]: OK， 我修改。

File: src/paddlefleet/transformer/moe/token_dispatcher.py [hunk 9]
```diff
@@ -1114,5 +1140,728 @@ def combine_postprocess(self, permutated_local_input_tokens: paddle.Tensor):
...
+        fn,
+        is_first_fwd=False,
+        fp8_combine_grad_handle=None,
+    ):
+        if fn is None:
+            raise ValueError(
+                "_AllGatherCombineAsync requires a non-None fn for overlap."
+            )
+        ctx.group = group
+        ctx.fp8_combine_grad_handle = fp8_combine_grad_handle
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/token_dispatcher.py:1491]: OK, 我修改。

File: tests/single_card_tests/ai_edited_test/moe/test_ai_allgather_dispatcher.py [hunk 10]
```diff
@@ -0,0 +1,984 @@
...
+        src = inspect.getsource(TransformerLayerWithOverlap.__init__)
+        return '"deepep"' in src and '"hybridep"' in src and "not in" in src
+
+    def test_guard_exists_in_source(self):
+        """The production __init__ has a not-in-('deepep','hybridep') guard."""
+        self.assertTrue(self._get_allowed_dispatchers_from_source())
+
+    def test_allgather_rejected_by_source(self):
+        """allgather is not in ('deepep', 'hybridep') → would be rejected."""
+        self.assertNotIn("allgather", ("deepep", "hybridep"))
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply tests/single_card_tests/ai_edited_test/moe/test_ai_allgather_dispatcher.py]: OK, 我修改。

File: tests/multi_card_tests/moe/test_allgather_dispatcher_ep.py [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+            mask,
+            topk_weights=tw,
+            topk_indices=ti,
+        )
+        x = paddle.randn([self.T_local * self.ep_size, self.hidden_size])
+        result, handle = dispatcher.token_dispatch(x, using_sonic_moe=True)
+        np.testing.assert_allclose(result.numpy(), x.numpy())
+        self.assertIsNone(handle)
+
+    def test_token_combine_reducescatter(self):
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply tests/multi_card_tests/moe/test_allgather_dispatcher_ep.py:282]: OK, 我修改。

File: tests/single_card_tests/ai_edited_test/moe/test_ai_allgather_dispatcher.py [hunk 12]
```diff
@@ -0,0 +1,1961 @@
...
+        self.assertEqual(fused.shape[0], T)
+
+
+class TestAllGatherGradFP8Async(unittest.TestCase):
+    """_all_gather_grad_fp8_async: lines 1361-1371."""
+
+    def test_quantize_and_launches_async_gather(self):
+        import paddlefleet.transformer.moe.token_dispatcher as td
+        from paddlefleet.transformer.moe.token_dispatcher import (
+            _all_gather_grad_fp8_async,
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply tests/single_card_tests/ai_edited_test/moe/test_ai_allgather_dispatcher.py]: ok，我修改。

**Review by ForFishes — APPROVED** [review]
LGTM

**Review by sneaxiy — APPROVED** [review]
